Multiplayer <feature

trombomart opened the pull request.

trombomart committed:
Add wallet as attribute of a player <organisation

Also you can get the wallet of a player in the model.

trombomart committed:
Test

trombomart committed:
Added wallet to View <feature

Commit:
Added buttons for preferences to menu < feature, user-interface

Commit:
Merge branch 'MultiPlayer' of https://github.com/hungnguyen94/BTrouble into MultiPlayer

trombomart committed:
Changed shopview for more players <feature

Only when versus is off!!

trombomart committed:
Two players in screen <feature

trombomart committed:
Goes on when one player is still alive <feature

trombomart committed:
Small fixes <bug

rubenwiersma commented on src/main/java/com/sem/btrouble/controller/Controller.java:
```diff
@@ -81,8 +82,8 @@ public void update(int delta) throws SlickException {
-
-            if (!player.isAlive()) {
+            
+            if (!anyLive()) {
```
spelling: anyLife()

rubenwiersma commented on src/main/java/com/sem/btrouble/controller/Controller.java:
```diff
@@ -92,7 +93,8 @@ public void update(int delta) throws SlickException {
         if (this.getTimers().getLevelTimeLeft() <= 0) {
-            fireEvent(new ControllerEvent(this, ControllerEvent.OUTOFTIME, "Out of time"));
+            fireEvent(new ControllerEvent(this, ControllerEvent.OUTOFTIME,
+                    "Out of time"));
```
line-break?

rubenwiersma commented on src/main/java/com/sem/btrouble/controller/Controller.java:
```diff
@@ -154,6 +156,29 @@ public void processInput(int delta) {
         }
+        if (Model.getPlayers().size() > 1) {
+            secondPlayer(delta);
+        }
```
Als secondPlayer in een aparte methode staat, kunnen de controls voor firstPlayer denk ik ook wel in een aparte methode. Misschien kan je zelfs een methode voor allebei de players maken waarbij je de player en keys als attributen doorgeeft. Als je wil, kan ik dat ook maken.

rubenwiersma commented on src/main/java/com/sem/btrouble/controller/Controller.java:
```diff
@@ -168,19 +193,29 @@ public void loseLife(Player player) {
         } else {
-            collisionHandler.removeCollidable(Model.getCurrentRoom().getCollidables());
+            collisionHandler.removeCollidable(Model.getCurrentRoom()
+                    .getCollidables());
```
line-break

rubenwiersma commented on src/main/java/com/sem/btrouble/controller/Controller.java:
```diff
@@ -168,19 +193,29 @@ public void loseLife(Player player) {
             // Hardcoded player 1
-            collisionHandler.removeCollidable(Model.getPlayers().get(0).getRopes());
+            collisionHandler.removeCollidable(Model.getPlayers().get(0)
+                    .getRopes());
```
line-break

rubenwiersma commented on src/main/java/com/sem/btrouble/controller/Controller.java:
```diff
@@ -168,19 +193,29 @@ public void loseLife(Player player) {
     private void restartRoom() {
-        fireEvent(new ControllerEvent(this, ControllerEvent.RESTARTROOM, "Room restarted"));
+        fireEvent(new ControllerEvent(this, ControllerEvent.RESTARTROOM,
+                "Room restarted"));
```
line-break

rubenwiersma commented on src/main/java/com/sem/btrouble/controller/Controller.java:
```diff
@@ -192,10 +227,12 @@ private void restartRoom() {
         if (!Model.getCurrentRoom().hasBubbles()) {
-            collisionHandler.removeCollidable(Model.getCurrentRoom().getCollidables());
+            collisionHandler.removeCollidable(Model.getCurrentRoom()
+                    .getCollidables());
```
line-break

rubenwiersma commented on src/main/java/com/sem/btrouble/controller/Controller.java:
```diff
@@ -192,10 +227,12 @@ private void restartRoom() {
             Model.getNextRoom();
             restartRoom();
-            //sbg.enterState(2, new FadeOutTransition(), new FadeInTransition());
+            // sbg.enterState(2, new FadeOutTransition(), new
```
commented code

rubenwiersma commented on src/main/java/com/sem/btrouble/view/GameView.java:
```diff
@@ -47,11 +50,13 @@ public void init(GameContainer gc, StateBasedGame sbg) throws SlickException {
+        ArrayList<Player> players = Model.getPlayers();
+        for(Player player: players) {
+            Wallet wallet = player.getWallet();
+            controller.registerObserver(wallet);
```
kan in één regel: controller.registerObserver(player.getWallet());

rubenwiersma commented on src/main/java/com/sem/btrouble/view/LostLevelView.java:
```diff
@@ -76,12 +82,14 @@ public void render(GameContainer gc, StateBasedGame sbg, Graphics graphics)
-        graphics.drawString("Wallet: " + GameView.getWallet().getValue(), 500, 660);
-        graphics.drawString("You died. Press enter to restart this Level", 350, 250);
+        graphics.drawString("You died. Press enter to restart this Level", 350,
+                250);
```
line-break

rubenwiersma commented on src/main/java/com/sem/btrouble/view/LostLevelView.java:
```diff
@@ -52,7 +57,8 @@ public void init(GameContainer gc, StateBasedGame sbg) throws SlickException {
      */
-    public void update(GameContainer gc, StateBasedGame sbg, int delta) throws SlickException {
+    public void update(GameContainer gc, StateBasedGame sbg, int delta)
+            throws SlickException {
```
line-break

rubenwiersma commented on src/main/java/com/sem/btrouble/view/LostLevelView.java:
```diff
@@ -33,8 +37,9 @@
-    public void init(GameContainer gc, StateBasedGame sbg) throws SlickException {
-        //Set objects to draw
+    public void init(GameContainer gc, StateBasedGame sbg)
+            throws SlickException {
```
line-break

rubenwiersma commented on src/main/java/com/sem/btrouble/view/LostLevelView.java:
```diff
@@ -76,12 +82,14 @@ public void render(GameContainer gc, StateBasedGame sbg, Graphics graphics)
 
-        SpriteSheet livesImage = new SpriteSheet("Sprites/lives_spritesheet.jpg", 381, 171);
+        SpriteSheet livesImage = new SpriteSheet(
+                "Sprites/lives_spritesheet.jpg", 381, 171);
```
line-break

rubenwiersma commented on src/main/java/com/sem/btrouble/view/LostLevelView.java:
```diff
@@ -89,16 +97,38 @@ public void render(GameContainer gc, StateBasedGame sbg, Graphics graphics)
+        ArrayList<Player> players = Model.getPlayers();
+        int sum = 0;
+
+        for (int i = 0; i < players.size(); i++) {
```
kan met foreach

rubenwiersma commented:
Alles gecheckt. Het lijkt erop dat Eclipse soms line-breaks neerzet die niet echt nodig zijn in de code. Ik heb volgens mij bij de meesten even line-break neergezet, zodat je die kan fixen.

trombomart committed:
Fixed code review issues <formatting

trombomart committed:
Add Lifes to screen and fixed little bug <feature, bug

Bugs left: Second player doesn't pick up an power up.
Shop screen has to be removed when versus is on.
When a player is hit and has still lifes left it has to come back
immediately with one life less.

trombomart committed:
Checkstyle fixes <formatting

hungnguyen94 commented on src/main/java/com/sem/btrouble/controller/Controller.java:
```diff
@@ -171,8 +192,18 @@ public void loseLife(Player player) {
             collisionHandler.removeCollidable(Model.getPlayers().get(0).getRopes());
+            if(SlickApp.multiplayer()) {
+                collisionHandler.removeCollidable(Model.getPlayers().get(1).getRopes());
+            }
```
You could use a for-each loop here instead.

hungnguyen94 commented on src/main/java/com/sem/btrouble/controller/Controller.java:
```diff
@@ -171,8 +192,18 @@ public void loseLife(Player player) {
             restartRoom();
             player.setAlive(true);
+            ArrayList<Player> players = Model.getPlayers();
+            for(Player otherPlayer: players) {
```
```
for(Player otherPlayer: Model.getPlayers()
```

might be a bit neater.

hungnguyen94 commented on src/main/java/com/sem/btrouble/controller/Controller.java:
```diff
@@ -289,24 +321,38 @@ public void removeObserver(LevelObserver observer) {
+     * Checks if there is a player in the room who is still alive.
+     * @return a boolean
+     */
+    public boolean anyLife() {
```
Instead of using a boolean, when assigning alive = true you could immediately return true.
Return false on the last line. 

```
for (Player player : Model.getPlayers() {
    if (player.isAlive()) {
        return true;
    }
}
return false;
```

hungnguyen94 commented on src/main/java/com/sem/btrouble/view/GameView.java:
```diff
@@ -47,11 +49,12 @@ public void init(GameContainer gc, StateBasedGame sbg) throws SlickException {
         soundObserver = new SoundObserver();
-        wallet = new Wallet();
+        ArrayList<Player> players = Model.getPlayers();
+        for(Player player: players) {
```
`for(Player player: Model.getPlayers())`

hungnguyen94 commented on src/main/java/com/sem/btrouble/view/View.java:
```diff
@@ -180,9 +192,24 @@ private void drawLives() throws SlickException {
-        graphics.drawString(score, 1090 - font.getWidth(score), 670);
+        for (int i = 0; i < players.size(); i++) {
+            String value = "Player " + (i + 1) + ": " 
+                   + Model.getWallet(Model.getPlayers().get(i)).getValue();
```
Doesn't the Player object itself contain a wallet? 
Why is this construction used?

hungnguyen94 commented on src/main/java/com/sem/btrouble/model/Model.java:
```diff
@@ -45,6 +45,15 @@ public static void init(int width, int height) {
+     * @param player from who you want the wallet
+     * @return the wallet
+     */
+    public static Wallet getWallet(Player player) {
```
Can't you get the wallet by using player.getWallet?

trombomart committed:
Some unit tests <testing

trombomart committed:
Code review fixes <formatting

trombomart committed:
Fixed Power up problem <bug

trombomart committed:
Removed failing test <bug

Commit:
Fixed merge conflicts < bug

Commit:
Fixed merge conflicts again < bug
